src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,138 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+import java.util.Objects;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+
+    private T[] elements;
+    private int size = 0;
+
+    public ArrayList() {
+        elements = (T[]) new Object[DEFAULT_CAPACITY];
+    }
+
+    private void ensureCapacity() {
+        T[] newElements = (T[]) new Object[elements.length * 2];
+        System.arraycopy(elements, 0, newElements, 0, size);
+        elements = newElements;
+    }
+
+    private boolean isFull() {
+        return size == elements.length;
+    }
+
+    private boolean isIndexValid(int index) {
+        return index < size && index >= 0;
+    }
+
+    private boolean isIndexValidForAdd(int index) {
+        return index <= size && index >= 0;
+    }
+
     @Override
     public void add(T value) {
-
+        if (isFull()) {
+            ensureCapacity();
+        }
+        elements[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (!isIndexValidForAdd(index)) {
+            throw new ArrayListIndexOutOfBoundsException(
+                    "You tried to add element to non exist index"
+            );
+        }
+        if (isFull()) {
+            ensureCapacity();
+        }
+        T[] newElements = (T[]) new Object[elements.length];
 
+        System.arraycopy(elements, 0, newElements, 0, index);
+        newElements[index] = value;
+        System.arraycopy(elements, index, newElements, index + 1, size - index);
+        size++;
+        elements = newElements;
     }
```

Comment:
The `add(T value, int index)` method is creating a new array every time it is called. This is not efficient and can be avoided. Instead of creating a new array, you can shift the elements to the right of the index to the right by one position, then add the new element at the index. This can be done using `System.arraycopy`.